### PR TITLE
Security second part

### PR DESCRIPTION
**Pull request by ShastkivRuslan**
Add Spring Security configuration and Role-based access control

- Integrate spring-boot-starter-security and create SecurityConfig with PasswordEncoder and SecurityFilterChain beans
- Implement UserDetails interface in User entity and create UserDetailsService implementation
- Make /api/auth/register endpoint public and restrict other endpoints based on user roles
- Create Role entity, RoleRepository, and establish Many-to-Many relationship between User and Role entities
- Add changeset to create roles in the database and assign roles to users
- Configure Swagger to remain accessible
- Update User entity to include roles, and use `@PreAuthorize` for role-based access control on endpoints

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -84,6 +84,11 @@
             <artifactId>springdoc-openapi-starter-webmvc-ui</artifactId>
             <version>2.3.0</version>
         </dependency>
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-security</artifactId>
+            <version>3.3.5</version>
```
Comment (Elena-Bruyako) [inline pom.xml]:
```suggestion
```

File: src/main/java/ruslan/shastkiv/bookstore/controller/BookController.java [hunk 2]
```diff
@@ -62,6 +81,7 @@ public BookDto createBook(@RequestBody @Valid CreateBookRequestDto bookDto) {
             description = "Update an existing book. "
                     + "Requires the book ID and the updated details in the request body."
     )
+    @PreAuthorize("hasRole('ADMIN')")
     @PutMapping("/{id}")
     public BookDto updateBook(@PathVariable Long id, @RequestBody CreateBookRequestDto requestDto) {
```
Comment (Elena-Bruyako) [inline src/main/java/ruslan/shastkiv/bookstore/controller/BookController.java]:
```suggestion
    public BookDto updateBook(@PathVariable Long id, @RequestBody @Valid CreateBookRequestDto requestDto) {
```

File: src/main/java/ruslan/shastkiv/bookstore/controller/BookController.java [hunk 3]
```diff
@@ -42,17 +44,34 @@ public List<BookDto> getAll(Pageable pageable) {
...
+            summary = "Search for books",
+            description = "Search for books by author, title, or price range. "
+                    + "If only one price is provided, it will be treated as the minimum price, "
+                    + "and results will be returned starting from that price. "
+                    + "If two prices are provided, they will be treated as "
+                    + "the minimum and maximum price range."
+    )
+    @PreAuthorize("hasRole('USER')")
+    @GetMapping("/search")
+    public List<BookDto> search(BookSearchParametersDto searchParametersDto, Pageable pageable) {
```
Comment (Elena-Bruyako) [inline src/main/java/ruslan/shastkiv/bookstore/controller/BookController.java]: check other Pageable
```suggestion
    public List<BookDto> search(BookSearchParametersDto searchParametersDto, @ParameterObject @PageableDefault Pageable pageable) {
```

File: src/main/java/ruslan/shastkiv/bookstore/model/Role.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+import jakarta.persistence.Entity;
+import jakarta.persistence.EnumType;
+import jakarta.persistence.Enumerated;
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+import jakarta.persistence.Table;
+import org.springframework.security.core.GrantedAuthority;
+
+@Entity
```
Comment (Elena-Bruyako) [inline src/main/java/ruslan/shastkiv/bookstore/model/Role.java:15]:
```suggestion
@Entity
@Getter
@Setter
```

File: src/main/java/ruslan/shastkiv/bookstore/model/User.java [hunk 5]
```diff
@@ -31,4 +42,50 @@ public class User {
...
 
     private String shippingAddress;
+
+    @ManyToMany
+    @JoinTable(
+            name = "users_roles",
+            joinColumns = @JoinColumn(name = "user_id"),
+            inverseJoinColumns = @JoinColumn(name = "role_id")
+    )
+    private Set<Role> roles;
```
Comment (Elena-Bruyako) [inline src/main/java/ruslan/shastkiv/bookstore/model/User.java]:
```suggestion
    private Set<Role> roles = new HashSet<>();
```

File: src/main/java/ruslan/shastkiv/bookstore/repository/user/UserRepository.java [hunk 6]
```diff
@@ -1,8 +1,14 @@
...
 import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.data.repository.query.Param;
 import ruslan.shastkiv.bookstore.model.User;
 
 public interface UserRepository extends JpaRepository<User, Long> {
     boolean existsByEmail(String email);
+
+    @Query("SELECT u FROM User u JOIN FETCH u.roles WHERE u.email = :email")
+    Optional<User> findUserWithRolesByEmail(@Param("email") String email);
```
Comment (Elena-Bruyako) [inline src/main/java/ruslan/shastkiv/bookstore/repository/user/UserRepository.java]:
```suggestion
    @EntityGraph(attributePaths = "roles")
    Optional<User> findByEmail(String email);
```

File: src/main/java/ruslan/shastkiv/bookstore/security/UserDetailsServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+import ruslan.shastkiv.bookstore.repository.user.UserRepository;
+
+@RequiredArgsConstructor
+@Service
+public class UserDetailsServiceImpl implements UserDetailsService {
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
+        return userRepository.findUserWithRolesByEmail(email).orElseThrow(
```
Comment (Elena-Bruyako) [inline src/main/java/ruslan/shastkiv/bookstore/security/UserDetailsServiceImpl.java]:
```suggestion
        return userRepository.findByEmail(email).orElseThrow(
```

File: src/main/java/ruslan/shastkiv/bookstore/security/UserDetailsServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+                () -> new UsernameNotFoundException("Can`t find user by email ["
```
Comment (Elena-Bruyako) [inline src/main/java/ruslan/shastkiv/bookstore/security/UserDetailsServiceImpl.java]:
```suggestion
                () -> new EntityNotFoundException("Can`t find user by email ["
```

File: src/main/java/ruslan/shastkiv/bookstore/service/user/UserServiceImpl.java [hunk 9]
```diff
@@ -22,7 +24,9 @@ public UserDto register(UserRegistrationRequestDto registrationRequestDto) {
                     + registrationRequestDto.email() + "] already registered!"
             );
         }
-        User savedUser = userRepository.save(userMapper.toModel(registrationRequestDto));
+        User userFromRequest = userMapper.toModel(registrationRequestDto);
+        userFromRequest.setPassword(passwordEncoder.encode(registrationRequestDto.password()));
+        User savedUser = userRepository.save(userFromRequest);
```
Comment (Elena-Bruyako) [inline src/main/java/ruslan/shastkiv/bookstore/service/user/UserServiceImpl.java]: set USER role to user

```suggestion
        User user = userMapper.toModel(registrationRequestDto);
        user.setPassword(passwordEncoder.encode(registrationRequestDto.password()));
        userRepository.save(user);
```

File: src/main/java/ruslan/shastkiv/bookstore/service/user/UserServiceImpl.java [hunk 10]
```diff
@@ -22,7 +24,9 @@ public UserDto register(UserRegistrationRequestDto registrationRequestDto) {
...
         return userMapper.toDto(savedUser);
```
Comment (Elena-Bruyako) [inline src/main/java/ruslan/shastkiv/bookstore/service/user/UserServiceImpl.java]:
```suggestion
        return userMapper.toDto(user);
```

File: src/main/resources/db/changelog/changes/02-create-users-table.yaml [hunk 11]
```diff
@@ -36,4 +36,4 @@ databaseChangeLog:
                     nullable: false
               - column:
                   name: shipping_address
-                  type: varchar(255)
\ No newline at end of file
+                  type: varchar(255)
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/02-create-users-table.yaml:39]: add isDeleted column

File: src/main/resources/db/changelog/changes/06-add-is_deleted-column.yaml [hunk 12]
```diff
@@ -0,0 +1,14 @@
+databaseChangeLog:
+  - changeSet:
+      id: add-is_deleted-column
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/06-add-is_deleted-column.yaml]: remove

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/ruslan/shastkiv/bookstore/controller/BookController.java [hunk 13]
```diff
@@ -32,8 +35,9 @@ public class BookController {
             summary = "Get all books",
             description = "Retrieve a list of all books. Supports pagination."
     )
+    @PreAuthorize("hasRole('USER')")
     @GetMapping
-    public List<BookDto> getAll(Pageable pageable) {
+    public List<BookDto> getAll(@ParameterObject @PageableDefault Pageable pageable) {
```
Comment (liliia-ponomarenko) [inline src/main/java/ruslan/shastkiv/bookstore/controller/BookController.java:40]: Better return Page instead List

File: src/main/resources/db/changelog/changes/02-create-users-table.yaml [hunk 14]
```diff
@@ -36,4 +36,10 @@ databaseChangeLog:
...
+              - column:
+                  name: is_deleted
+                  type: tinyint(1)
+                  defaultValueBoolean: false
+                  constraints:
+                    nullable: false
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/changes/02-create-users-table.yaml:45]: Better create a new changeSet. On a real project, you would get a checksum mismatch error

**Review by liliia-ponomarenko — APPROVED** [review]
Great 🔥 A few recommendations ;)

**Pull request merged**